rasa/graph_components/providers/nlu_training_data_provider.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import annotations
+from typing import Dict, Text, Any, Optional
+import os
+from rasa.engine.graph import GraphComponent, ExecutionContext
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+from rasa.shared.importers.importer import TrainingDataImporter
+from rasa.shared.nlu.training_data.training_data import (
+    TrainingData,
+    DEFAULT_TRAINING_DATA_OUTPUT_PATH,
+)
+from rasa.shared.nlu.training_data.loading import load_data
+
+
+class NLUTrainingDataProvider(GraphComponent):
+    """Provides NLU training data during training and inference time."""
+
+    def __init__(
+        self,
+        model_storage: ModelStorage,
+        resource: Resource,
+        training_data: Optional[TrainingData] = None,
+    ) -> None:
+        """Creates NLU training data provider."""
+        self._model_storage = model_storage
+        self._resource = resource
+        self._training_data = training_data
+
+    @classmethod
+    def get_default_config(cls) -> Dict[Text, Any]:
+        return {"language": "en", "persist": False}
+
+    @classmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> NLUTrainingDataProvider:
+        """Creates component (see parent class for full docstring)."""
+        return cls(model_storage, resource)
+
+    @classmethod
+    def load(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+        **kwargs: Any,
+    ) -> NLUTrainingDataProvider:
+        """Creates provider using a persisted data."""
+        with model_storage.read_from(resource) as resource_directory:
+            training_data = load_data(
+                resource_name=os.path.join(
+                    str(resource_directory), DEFAULT_TRAINING_DATA_OUTPUT_PATH
+                ),
+                language=config["language"],
+            )
+        return cls(model_storage, resource, training_data)
+
+    def _persist(self, training_data: TrainingData) -> None:
+        """Persists NLU training data to model storage."""
+        with self._model_storage.write_to(self._resource) as resource_directory:
+            training_data.persist(
+                dir_name=str(resource_directory),
+                filename=DEFAULT_TRAINING_DATA_OUTPUT_PATH,
+            )
```

Comment:
I used the constant `DEFAULT_TRAINING_DATA_OUTPUT_PATH` from `training_data.py` instead of hardcoding the nlu file name the way `DomainProvider` [does](https://github.com/RasaHQ/rasa/blob/33267a951f924ed17250a8e3b1c4f136b1670c16/rasa/graph_components/providers/domain_provider.py#L53). Is this okay, or should I hardcode it to be consistent?

I noticed that the interface for [`persist` in `TrainingData`](https://github.com/RasaHQ/rasa/blob/33267a951f924ed17250a8e3b1c4f136b1670c16/rasa/shared/nlu/training_data/training_data.py#L431) is differnt to [the one for `Domain`](https://github.com/RasaHQ/rasa/blob/33267a951f924ed17250a8e3b1c4f136b1670c16/rasa/shared/core/domain.py#L1436). Is this okay?

Reply:
> hardcoding the nlu file name the way DomainProvider does. Is this okay, or should I hardcode it to be consistent?
that's ok  - the names don't really matter (they only need to be unique within a "resource" ^^)

> interface ... is different
that is ok as well - domain and training data don't have to implement the method in the same way (if there was a common "persistable" interface that they implemented, then they should look the same) 

(@joejuzl , we could also create a cacheable version of TrainingData and let the caching mechanism take care of persisting, but that would mean that we cannot make persisting optional as we can via this approach here where we can switch that on/off via the config... 🤔  Which is not what we want, right?)

Reply:
We only persist so that the data is in the model package at the end. I think any caching would be separate from this.